src/transaction/flashback.c

```diff
@@ -0,0 +1,418 @@
+  int supplement_alloc_length = 0;
+
+  log_page_p = (LOG_PAGE *) PTR_ALIGN (log_pgbuf, MAX_ALIGNMENT);
+
+  assert (!LSA_ISNULL (&context->end_lsa));
+
+  LSA_COPY (&process_lsa, &(context->end_lsa));
+
+  /*fetch log page */
+  if (logpb_fetch_page (thread_p, &process_lsa, LOG_CS_SAFE_READER, log_page_p) != NO_ERROR)
+    {
+      goto exit;
+    }
+
+  while (LSA_GE (&process_lsa, &context->start_lsa))
+    {
+      LSA_COPY (&cur_log_rec_lsa, &process_lsa);
+      log_rec_header = LOG_GET_LOG_RECORD_HEADER (log_page_p, &process_lsa);
+
+      log_type = log_rec_header->type;
+      trid = log_rec_header->trid;
+
+      LSA_COPY (&next_log_rec_lsa, &log_rec_header->back_lsa);
+
+      if (LSA_ISNULL (&log_rec_header->prev_tranlsa))
+	{
+	  FLASHBACK_CHECK_AND_GET_SUMMARY (context->summary_list, trid, summary_entry);
+	  if (summary_entry != NULL)
+	    {
+	      /* current time is updated whenever a log record with a time value is encountered.
+	       * this function reads the log records in reverse time order,
+	       * so the current_time represents a time in the future than the log being read now
+	       * Since LOG_DUMMY_HA_SERVER_STATE is logged every second, start_time is set to 1 second before the current_time */
+
+	      summary_entry->start_time = current_time - 1;
+	      LSA_COPY (&summary_entry->start_lsa, &cur_log_rec_lsa);
+	    }
+	}
+
+      LOG_READ_ADD_ALIGN (thread_p, sizeof (*log_rec_header), &process_lsa, log_page_p);
+
+      switch (log_type)
+	{
+	case LOG_COMMIT:
+	  {
+	    LOG_REC_DONETIME *donetime;
+
+	    if (context->num_summary == FLASHBACK_MAX_SUMMARY)
+	      {
+		error = ER_FLASHBACK_TOO_MANY_SUMMARY;
+		goto exit;
+	      }
+
+	    LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*donetime), &process_lsa, log_page_p);
+
+	    donetime = (LOG_REC_DONETIME *) (log_page_p->area + process_lsa.offset);
+
+	    /* create summary info entry, and register it to the summary list */
+	    if ((error = flashback_create_summary_entry (thread_p, &summary_entry)) != NO_ERROR)
+	      {
+		goto exit;
+	      }
+
+	    summary_entry->trid = trid;
+	    LSA_COPY (&summary_entry->end_lsa, &cur_log_rec_lsa);
+	    summary_entry->end_time = donetime->at_time;
+
+        // *INDENT-OFF*
+	context->summary_list.insert (std::make_pair (trid, summary_entry));
+        // *INDENT-ON*
+	    current_time = donetime->at_time;
+	    context->num_summary++;
+	    break;
+	  }
+	case LOG_DUMMY_HA_SERVER_STATE:
+	  {
+	    LOG_REC_HA_SERVER_STATE *ha_dummy;
+
+	    LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*ha_dummy), &process_lsa, log_page_p);
+
+	    ha_dummy = (LOG_REC_HA_SERVER_STATE *) (log_page_p->area + process_lsa.offset);
+
+	    current_time = ha_dummy->at_time;
+
+	    break;
+	  }
+	case LOG_SUPPLEMENTAL_INFO:
+	  {
+	    LOG_REC_SUPPLEMENT *supplement;
+	    int supplement_length;
+	    SUPPLEMENT_REC_TYPE rec_type;
+	    RECDES supp_recdes = RECDES_INITIALIZER;
+
+	    OID classoid;
+
```

Comment:
At this point, it seems that we are able to skip the following works if `FLASHBACK_CHECK_AND_GET_SUMMARY (context->summary_list, trid, summary_entry)` return NULL. I remember that it is one of the reasons we traverse the log backward.